lib/autoBlock.js

```diff
@@ -0,0 +1,215 @@
+  }
+
+  var status = _.get(error, 'data.status')
+  if (error && error.message && status) {
+    error.message = (httpStatus[status] || status) + `: ${error.message}`
+  }
+
+  if (error) {
+    config.onFailure(error, data)
+  } else {
+    config.onSuccess(data.response, data)
+  }
+
+  var done = config.done
+  if (done) {
+    // Lambda will automatically retry if you send the error back
+    if (!_.get(error, 'data.retry', true)) {
+      error = null
+    }
+    process.nextTick(() => done(error, data.response))
+  }
+}
+
+// Build out an async.auto block using our special notation
+function buildBlock (block, config) {
+  var autoBlock = _.transform(block, (asyncBlock, step, key) => {
+    // If no name was provided, use the key. This is used during logging.
+    step.name = step.name || key
+
+    // Figure out which steps need to come before this one
+    // - anything in .after
+    // - the first path segement in each .with
+    // - the first path segement in each .when
+    var deps = (step.after || []).slice()
+
+    var withDeps = (step.with || [])
+    if (_.isString(withDeps)) {
+      withDeps = [withDeps]
+    }
+    withDeps = _.map(withDeps, (value) => _.first(_.split(value, '.')))
+
+    var whenDeps = (step.when || [])
+    if (_.isString(whenDeps)) {
+      whenDeps = [whenDeps]
+    }
+    whenDeps = _.map(whenDeps, (value) => {
+      var dep = _.first(_.split(value, '.'))
+      if (_.first(dep) === '!') {
+        return dep.substring(1)
+      }
+
+      return dep
+    })
+
+    deps = _.union(deps, withDeps, whenDeps)
+
+    // Build the async callback. This takes two forms: If no dependencies, no results paramater is included.
+    var asyncFunc
+    if (deps.length) {
+      asyncFunc = (results, cb) => runStep(results, cb, step, config)
+    } else {
+      asyncFunc = (cb) => runStep(null, cb, step, config)
+    }
+
+    // async.auto takes the callback as the final element in the depedencies array
+    deps.push(asyncFunc)
+
+    asyncBlock[key] = deps
+  }, {})
+
+  return autoBlock
+}
+
+// Run an individual step
+function runStep (results, callback, step, config) {
+  // .value simply returns
+  if (step.value) {
+    return callback(null, step.value)
+  }
+
+  // Only run this step if all .when values resolve truthy. (But don't error; just skip the step if a check fails.)
+  var checks = step.when || []
+  if (!_.isArray(checks)) {
+    checks = [checks]
+  }
+  var valid = _.every(checks, (check) => {
+    if (_.first(check) === '!') {
+      return !_.get(results, check.substring(1))
+    }
+
+    return _.get(results, check)
+  })
+
+  if (!valid) {
+    return callback()
+  }
+
+  // Run the inner function using the wrappers
+  runStepFunc(results, callback, step, config)
+}
+
+// Log and run the step with logging
+function runStepFunc (results, callback, step, config) {
+  var func = step.func
+  var start = Date.now()
+
+  config.onStartStep(step.name, config.data)
```

Comment:
The start time will include the external `onStartStep` code, is that by design?

Also, shouldn't the `on` functions by async?